### PR TITLE
Add Makefile project for RISC-V QEMU virtpc

### PR DESCRIPTION
**Pull request by katsuster**
Description
-----------
This patch adds simple demo directory for QEMU virtpc machine.

A demo just prints Tx/Rx message of queue to serial port, use no
other hardware and use only primary core (currently hart0).
Other cores are simply going to wfi state and execute nothing else.

Test Steps
-----------
  $ make
  $ qemu-system-riscv32 -nographic -machine virt -net none \
    -chardev stdio,id=con,mux=on -serial chardev:con \
    -mon chardev=con,mode=readline -bios none \
    -kernel demo.elf -smp cpus=2

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/FreeRTOSConfig.h [hunk 1]
```diff
@@ -0,0 +1,104 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/FreeRTOSConfig.h:2]: we are 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/FreeRTOSConfig.h [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/FreeRTOSConfig.h:23]: https://www.FreeRTOS.org
 https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+CC=$(CROSS)gcc
+OBJCOPY=$(CROSS)objcopy
+ARCH=$(CROSS)ar
+
+RTOS_SOURCE_DIR=../../Source
+DEMO_SOURCE_DIR=../Common/Minimal
+LIBWRAP_SOURCE_DIR=./freedom-e-sdk/libwrap
+
+CPPFLAGS = \
+	-g -O2 -Wall \
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile:11]: could you please add a debug flag (DEBUG=1) to pass to the makefile to add debugging symbols with   (-Og -ggdb3)

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile [hunk 4]
```diff
@@ -0,0 +1,53 @@
+CROSS=riscv64-unknown-elf-
+CC=$(CROSS)gcc
+OBJCOPY=$(CROSS)objcopy
+ARCH=$(CROSS)ar
+
+RTOS_SOURCE_DIR=../../Source
+DEMO_SOURCE_DIR=../Common/Minimal
+LIBWRAP_SOURCE_DIR=./freedom-e-sdk/libwrap
+
+CPPFLAGS = \
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile:10]: most of these are CFLAGS... it doesn't matter just for completeness

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	$(RTOS_SOURCE_DIR)/timers.c \
+	$(RTOS_SOURCE_DIR)/portable/MemMang/heap_4.c \
+	$(RTOS_SOURCE_DIR)/portable/GCC/RISC-V/port.c
+
+ASMS = start.S \
+	$(RTOS_SOURCE_DIR)/portable/GCC/RISC-V/portASM.S
+
+OBJS = $(SRCS:.c=.o) $(ASMS:.S=.o)
+
+demo.elf: $(OBJS) fake_rom.lds Makefile
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile:49]: We use a common name which is RTOSDemo.axf

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	$(CC) $(CFLAGS) $(LDFLAGS) $(OBJS) -o $@
+
+clean:
+	rm -rf demo.elf $(OBJS)
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile:53]: we prefer out of source build, so you can just delete the build directory

also missing a step to resolve header dependencies automatically for recompilation

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c [hunk 7]
```diff
@@ -0,0 +1,231 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:23]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+	}
+}
+
+/*-----------------------------------------------------------*/
+
+int main( void )
+{
+	vSendString( "Hello FreeRTOS!" );
+
+	/* Create the queue. */
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/main.c:153]: could you please separate the main file from the file that actually do the demo
this way if someone wants to add more demos, their job will be easier, by just calling the appropriate function 
```c
#if defined BLINKY
    main_blinky()
#elif defined NETWORKING
    main_networking()
#endif 
```
for example

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.c:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.c [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.c:23]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.h:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.h [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/ns16550.h:23]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-reg.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-reg.h:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-reg.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-reg.h:23]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.c [hunk 16]
```diff
@@ -0,0 +1,65 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.c:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.c [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.c:22]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.h [hunk 18]
```diff
@@ -0,0 +1,55 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.h:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.h [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/riscv-virt.h:23]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/start.S [hunk 20]
```diff
@@ -0,0 +1,85 @@
+/*
+ * FreeRTOS Kernel V10.4.0
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/start.S:2]: 10.4.1

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/start.S [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/start.S:23]: https://www.FreeRTOS.org
https://www.github.com/FreeRTOS

File: FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+	$(RTOS_SOURCE_DIR)/portable/MemMang/heap_4.c \
+	$(RTOS_SOURCE_DIR)/portable/GCC/RISC-V/port.c
+
+ASMS = start.S \
+	$(RTOS_SOURCE_DIR)/portable/GCC/RISC-V/portASM.S
+
+OBJS = $(SRCS:.c=.o) $(ASMS:.S=.o)
+
+demo.elf: $(OBJS) fake_rom.lds Makefile
+	$(CC) $(CFLAGS) $(LDFLAGS) $(OBJS) -o $@
```
Comment (alfred2g) [inline FreeRTOS/Demo/RISC-V-Qemu-virt_GCC/Makefile:50]: no need for CFLAGS in linking step

**Review by alfred2g — COMMENTED** [review]
Could you also add a readme file Readme.md to describe the requirements and exact steps (compile/build, run, possibly debug) needed to make this work,
including links to where to download the compiler and qemu

**Comment by katsuster** [comment]
@alfred2g , Hello Alfred,
Thanks a lot for your review.

As you see, I added 4 more patches after previous patch to cover your review points.
If it seems acceptable, I will squash all patches into one.

**Comment by alfred2g** [comment]
> @alfred2g , Hello Alfred,
> Thanks a lot for your review.
> 
> As you see, I added 4 more patches after previous patch to cover your review points.
> If it seems acceptable, I will squash all patches into one.

NP!
no need to squash them yourself, github will do it automatically.

**Review by alfred2g — APPROVED** [review]

**Review by dan4thewin — APPROVED** [review]

**Pull request merged**